docs/dev/pagination-v2.md

```diff
@@ -0,0 +1,173 @@
+# Pagination in v2 Engine
+
+<!-- Why is pagination necessary? -->
+Pagination allows a SQL plugin client to retrieve arbitrarily large
+results sets one subset at a time.
+
+A cursor is a SQL abstraction for pagination. A client can open a cursor, retrieve a subset of data given a cursor and close a cursor.
+<!-- What applications will use pagination? -->
+
+Currently, SQL plugin does not provide SQL cursor syntax. However, the SQL REST endpoint can return result a page at a time.
+
+JDBC and ODBC drivers are the main users of cursor REST API.
+
+<!-- How will applications use pagination? -->
+
+# Scope
+<!-- What scenarios are excluded from this discussion? Why? -->
+
+The primary goal is to reach feature parity with the v1 engine. v1 engine suppors pagination only for simple `SELECT` queries with optional `WHERE` or `ORDER BY` clauses.
+
+## Compatibility with v1 engine
+<!-- How can one tell that v2 engine pagination is compatible with v1? -->
+
+The REST API will remain the same. However, v1 cursors are a thin wrapper over OpenSearch's scroll API.
+This is not possible with v2 engine because it may perform some processing in-memory
+
+# Behaviour Specification
+## Default Scenario
+
+A client requests a paginated response, retrieves data a subset at a time, and closes the cursor.
+
+```mermaid
+flowchart LR
+    start[query, fetch_size > 0]
+    fetch_some[get page]
+    close[close cursor]
+    start --> fetch_some 
+    fetch_some --> |more data | fetch_some 
+    fetch_some --> close
+    
+```
+Each subset response also includes a possibly new value for `cursor_id`. Client must ensure
+to pass the correct cursor id with each page request.
+
+
+### Client submits request with non-existing cursor_id 
+
+A client will receive an error response if it sends a request with a cursor_id that SQL plugin does not maintain.
+
+This can happen when client requests a paginated response but the cursor times out before all the data is retrieved.
+
+Response will include `error` and `status` fields.
+
+## Unable to create a new cursor [new]
+SQL plugin will have a limit on number of open cursors.
```

Comment:
How?  Where is this specified?  What happens if we exceed this number?  Is there a way to manage this number (manually close cursors)?  What about the timeout?  Can we specify this for the SQL plugin API when we create cursors or is it always defaulted to the default timeout (can this be configured).